core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala

```diff
@@ -35,16 +37,28 @@ case class ExecutorCacheTaskLocation(override val host: String, executorId: Stri
 
 /**
  * A location on a host.
+ * The host names will be resolved to IP addresses in case SPARK_LOCAL_HOSTNAME is unset
+ * to fix the task locality levels always being 'ANY' issue mentioned in SPARK-10149.
  */
 private [spark] case class HostTaskLocation(override val host: String) extends TaskLocation {
-  override def toString: String = host
+  override def toString: String = if ( TaskLocation.sparkLocalHostname == None ) {
+    InetAddress.getByName(host).getHostAddress
+  } else {
+    host
+  }
 }
 
 /**
  * A location on a host that is cached by HDFS.
+ * The host names will be resolved to IP addresses in case SPARK_LOCAL_HOSTNAME is unset
+ * to fix the task locality levels always being 'ANY' issue mentioned in SPARK-10149.
  */
 private [spark] case class HDFSCacheTaskLocation(override val host: String) extends TaskLocation {
-  override def toString: String = TaskLocation.inMemoryLocationTag + host
+  override def toString: String = if ( TaskLocation.sparkLocalHostname == None ) {
+    TaskLocation.inMemoryLocationTag + InetAddress.getByName(host).getHostAddress
```

Comment:
better handle `InetAddress.getByName()` failure conditions here. It is entirely possible for a host to not know its own name if configured badly enough.